[pp] pipelining + dcp unit test

H-Huang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #139633

Currently there aren't any unit tests for PP and DCP, this unit test could be useful for quick experimentation in issues like (https://github.com/pytorch/torchtitan/issues/474).

`python test/distributed/_composable/test_composability/test_pp_composability.py -k test_pp_and_dcp`

cc @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

H-Huang committed:
[pp] pipelining + dcp unit test

[ghstack-poisoned]

wconstab commented on test/distributed/_composable/test_composability/test_pp_composability.py:
```diff
@@ -253,6 +265,112 @@ def build_stage(stage_idx, num_stages):
+                sd,
+                storage_reader=FileSystemReader(checkpoint_dir),
+                planner=_EmptyStateDictLoadPlanner(),
+                no_dist=True,
```
i forgot what this flag means?  (no_dist)

H-Huang replied:
looks like it is for allowing this to work even if distributed is not initialized, so this is not necessary and I am removing

wconstab commented on test/distributed/_composable/test_composability/test_pp_composability.py:
```diff
@@ -253,6 +265,112 @@ def build_stage(stage_idx, num_stages):
+                    sd_param_names,
+                    f"Parameter name '{param_name}' not found in state_dict.",
+                )
+            # Verify that the parameter layers include the start_index and end_index
```
well technically they would include the start_index but exclude the end_index right?

wconstab replied:
but why do we even need this assert? this isn't really testing DCP, its just testing whether you wrote the test correctly which i guess is fine to test but maybe not necessary?

wconstab commented on test/distributed/_composable/test_composability/test_pp_composability.py:
```diff
@@ -253,6 +265,112 @@ def build_stage(stage_idx, num_stages):
+                planner=_EmptyStateDictLoadPlanner(),
+                no_dist=True,
+            )
+            # Check parameter names in sd and compare with pp_model
```
do you think it is worth it to test optimizer keys as well? I assume it might be more complicated to write the test so maybe not desirable, not sure if it is worth the effort

fegin replied:
We may want to test when optimizer state_dict keys can conflict within a rank.

wconstab reviewed (approved):
good idea! thanks for adding

fegin commented on test/distributed/_composable/test_composability/test_pp_composability.py:
```diff
@@ -253,6 +265,112 @@ def build_stage(stage_idx, num_stages):
+        opt = torch.optim.Adam(pp_model.parameters(), lr=0.1)
+
+        # Manually create a temporary directory
+        # Create a temporary directory on rank 0 and broadcast its path
```
This can be tricky sometimes. Like rank0 creates the folder but is not seen by other ranks on the same machine. I would suggest to use `from torch.testing._internal.distributed.checkpoint_utils import with_temp_dir`.

wz337 replied:
You can directly use `self.temp_dir` when using @with_temp_dir. See example here: https://github.com/pytorch/pytorch/blob/fdfd4c50bad8b7d12416d45233ad990c45cf7ef9/test/distributed/checkpoint/test_format_utils.py#L50

H-Huang replied:
thanks, updated to use `@with_temp_dir`

wz337 commented on test/distributed/_composable/test_composability/test_pp_composability.py:
```diff
@@ -253,6 +265,112 @@ def build_stage(stage_idx, num_stages):
+        finally:
+            # Clean up the temporary directory on rank 0
+            torch.distributed.barrier()
+            if self.rank == 0:
```
When you switch to use @with_temp_dir, it would remove the directory for you. Code pointer: https://github.com/pytorch/pytorch/blob/fdfd4c50bad8b7d12416d45233ad990c45cf7ef9/torch/testing/_internal/distributed/checkpoint_utils.py#L46-L49

H-Huang committed:
Update on "[pp] pipelining + dcp unit test"


Currently there aren't any unit tests for PP and DCP, this unit test could be useful for quick experimentation in issues like (https://github.com/pytorch/torchtitan/issues/474).

`python test/distributed/_composable/test_composability/test_pp_composability.py -k test_pp_and_dcp`

cc awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

H-Huang commented:
@pytorchbot merge
